src/firecracker/examples/uffd/uffd_utils.rs

```diff
@@ -162,6 +162,110 @@ impl UffdPfHandler {
     }
 }
 
+#[derive(Debug)]
+pub struct Runtime {
+    stream: UnixStream,
+    backing_file: File,
+    backing_memory: *mut u8,
+    backing_memory_size: usize,
+    uffds: HashMap<i32, UffdPfHandler>,
+}
+
+impl Runtime {
+    pub fn new(stream: UnixStream, backing_file: File) -> Self {
+        let file_meta = backing_file
+            .metadata()
+            .expect("can not get backing file metadata");
+        let backing_memory_size = file_meta.len() as usize;
+        // # Safety:
+        // File size and fd are valid
+        let ret = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                backing_memory_size,
+                libc::PROT_READ,
+                libc::MAP_PRIVATE,
+                backing_file.as_raw_fd(),
+                0,
+            )
+        };
+        if ret == libc::MAP_FAILED {
+            panic!("mmap on backing file failed");
+        }
+
+        Self {
+            stream,
+            backing_file,
+            backing_memory: ret.cast(),
+            backing_memory_size,
+            uffds: HashMap::default(),
+        }
+    }
+
+    /// Polls the `UnixStream` and UFFD fds in a loop.
+    /// When stream is polled, new uffd is retrieved.
+    /// When uffd is polled, page fault is handled by
+    /// calling `pf_event_dispatch` with corresponding
+    /// uffd object passed in.
+    pub fn run(&mut self, pf_event_dispatch: impl Fn(&mut UffdPfHandler)) {
+        let mut pollfds = vec![];
+
+        // Poll the stream for incoming uffds
+        pollfds.push(libc::pollfd {
+            fd: self.stream.as_raw_fd(),
+            events: libc::POLLIN,
+            revents: 0,
+        });
+
+        // We can skip polling on stream fd if
+        // the connection is closed.
+        let mut skip_stream: usize = 0;
+        loop {
+            let pollfd_ptr = pollfds[skip_stream..].as_mut_ptr();
+            let pollfd_size = pollfds[skip_stream..].len() as u64;
+
+            // # Safety:
+            // Pollfds vector is valid
+            let mut nready = unsafe { libc::poll(pollfd_ptr, pollfd_size, -1) };
```

Comment:
Nit: It is always preferable to use `epoll` over `poll`.

From https://man7.org/linux/man-pages/man7/epoll.7.html

> By contrast, when used as a level-triggered interface (the
> default, when EPOLLET is not specified), epoll is simply a faster
> [poll(2)](https://man7.org/linux/man-pages/man2/poll.2.html), and can be used wherever the latter is used since it
> shares the same semantics.

There is also a nice nix interface for epoll (I'm certainly not biased since I wrote it 😊) https://docs.rs/nix/latest/nix/sys/epoll/struct.Epoll.html